File: tools/richos-service/companion-windows/src/RichOSCompanionCore/SessionWriter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace RichOSCompanionCore
{
    /// <summary>
    /// Orchestrates one capture session's on-disk contract directory (architecture §6.1 never-silent). The
    /// exact peer of the macOS companion's <c>SessionWriter.swift</c>.
    ///
    /// Lifecycle:
    ///   1. <see cref="Start"/> — create the dir, write <c>session.json</c> (status "open", empty
    ///      audio) IMMEDIATELY, open the WAV. A call that captures nothing therefore leaves a loud
    ///      open/empty session on disk — an anomaly the pipeline's reconcile guard flags — never an
    ///      absence nobody notices.
    ///   2. <see cref="PushMic"/> / <see cref="PushSystem"/> — real-time callbacks feed mono float frames.
    ///   3. <see cref="Pump"/> — mix aligned frames (with failover), append to WAV, accrue byte growth.
    ///   4. <see cref="WriteHealthTick"/> — one <c>health.ndjson</c> row per second.
    ///   5. <see cref="Close"/> — flush WAV, rewrite <c>session.json</c> (status "closed", populated).
    ///      The closed status is what triggers P1's pipeline.
    /// </summary>
    public sealed class SessionWriter : IDisposable
    {
        public string Dir { get; }
        public SessionContract.Params Params { get; }
        private readonly Func<long> _clock;

        private WavWriter? _wav;
        private readonly RingBuffer _micRing = new RingBuffer();
        private readonly RingBuffer _sysRing = new RingBuffer();

        private long? _firstAudioMs;
        private long? _lastAudioMs;
        private int _lastBytesTotal;
        private readonly SessionContract.HealthSummary _health = new SessionContract.HealthSummary();
        private FileStream? _healthHandle;

        /// <summary>Failover threshold: a source silent this long while the other is live trips silence-fill+alarm.</summary>
        public long StarveThresholdMs { get; set; } = 15_000;

        public bool Closed { get; private set; }
        public MixDecision? LastDecision { get; private set; }

        public SessionWriter(string rootZone, SessionContract.Params p, Func<long> clock)
        {
            Dir = System.IO.Path.Combine(rootZone, p.SessionId);
            Params = p;
            _clock = clock;
        }

        /// <summary>Write <c>session.json</c> (open) + open the WAV part. Call once.</summary>
        public void Start()
        {
            Directory.CreateDirectory(Dir);
            WriteSessionJson("open", null, Array.Empty<SessionContract.AudioPart>());
            string wavPath = System.IO.Path.Combine(Dir, SessionContract.AudioPartFile(0));
            _wav = new WavWriter(wavPath, Params.SampleRate, channels: 2, bitsPerSample: 16);
            string healthPath = System.IO.Path.Combine(Dir, "health.ndjson");
            _healthHandle = new FileStream(healthPath, FileMode.Create, FileAccess.Write, FileShare.Read);
        }

        /// <summary>
        /// Rewrite <c>session.json</c> in the OPEN state after mutating <see cref="Params"/> — used by
        /// the live path to correct <c>capture.method</c>/<c>captureTarget</c> once the WASAPI engine
        /// has resolved whether it got process loopback or fell back to system loopback. Never touches
        /// the WAV, so the never-silent open record simply gets more accurate.
        /// </summary>
        public void RewriteOpenSessionJson()
        {
            if (Closed) return;
            WriteSessionJson("open", null, Array.Empty<SessionContract.AudioPart>());
        }

        public void PushMic(IReadOnlyList<float> mono) => _micRing.Append(mono, _clock());
        public void PushSystem(IReadOnlyList<float> mono) => _sysRing.Append(mono, _clock());

        /// <summary>Mix one block of aligned frames to the WAV. Returns bytes written this pump.</summary>
        public int Pump()
        {
            if (_wav == null || Closed) return 0;
            long now = _clock();
            long micSilent = _micRing.LastAppend() == 0 ? 0 : now - _micRing.LastAppend();
            long sysSilent = _sysRing.LastAppend() == 0 ? 0 : now - _sysRing.LastAppend();
            var decision = MixDecision.Decide(
                _micRing.Count, _sysRing.Count, micSilent, sysSilent, StarveThresholdMs);
            LastDecision = decision;
            if (decision.Frames <= 0) return 0;

            float[] mic = decision.UseMic ? _micRing.Take(decision.Frames) : new float[decision.Frames];
            float[] sys = decision.UseSystem ? _sysRing.Take(decision.Frames) : new float[decision.Frames];
            // Guard alignment even under failover silence-fill.
            int n = Math.Min(mic.Length, sys.Length);
            if (n < mic.Length) Array.Resize(ref mic, n);
            if (n < sys.Length) Array.Resize(ref sys, n);
            short[] interleaved = ChannelMixer.InterleaveToInt16(mic, sys);
            byte[] bytes = ChannelMixer.Int16LeBytes(interleaved);
            _wav.Append(bytes);
            if (_firstAudioMs == null) _firstAudioMs = now;
            _lastAudioMs = now;
            return bytes.Length;
        }

        /// <summary>Flush the WAV to disk — the &lt;= one-flush-window crash guarantee.</summary>
        public void Flush() => _wav?.Flush();

        /// <summary>Total bytes on disk for the audio part so far — the watchdog's ground-truth signal.</summary>
        public int PumpedBytesTotal => _wav?.TotalBytesOnDisk ?? 0;

        /// <summary>
        /// Emit one health row + accrue the rolled-up summary from precomputed per-channel levels
        /// (peak + RMS over the last second). Kept level-based (not block-based) so the live capture can
        /// meter cheaply in its audio callbacks without re-buffering a second of samples.
        /// </summary>
        public void WriteHealthTick(
            double micPeak, double micRms, double sysPeak, double sysRms, bool tapRunning, bool micRunning)
        {
            long now = _clock();
            int bytesTotal = _wav?.TotalBytesOnDisk ?? 0;
            int delta = bytesTotal - _lastBytesTotal;
            _lastBytesTotal = bytesTotal;

            var problems = new List<string>();
            string level = "green";
            // Never-silent signals (§6.2): stalled byte growth, and failover-tripped starvation.
            if (_health.Heartbeats > 20 && delta == 0)
            {
                problems.Add("no audio bytes written in the last second");
                level = "amber";
            }
            if (LastDecision.HasValue && LastDecision.Value.MicStarved)
            {
                problems.Add("microphone source stalled — silence-filling LEFT (loopback-only failover the other way)");
                level = "red";
            }
            if (LastDecision.HasValue && LastDecision.Value.SystemStarved)
            {
                problems.Add("system loopback stalled — silence-filling RIGHT (device change or render endpoint went idle)");
                level = "red";
            }
            if (!tapRunning) { problems.Add("system loopback capture client not running"); level = "red"; }
            if (!micRunning) { problems.Add("microphone capture client not running"); level = "red"; }

            var sample = new SessionContract.HealthSample
            {
                SessionId = Params.SessionId,
                T = now,
                MicRms = micPeak,
                SysRms = sysPeak,
                MicRmsMean = micRms,
                SysRmsMean = sysRms,
                BytesTotal = bytesTotal,
                BytesDelta = delta,
                Part = 0,
                TapRunning = tapRunning,
                MicRunning = micRunning,
                Level = level,
                Problems = problems,
            };
            if (_healthHandle != null)
            {
                byte[] line = Encoding.UTF8.GetBytes(sample.ToJson().Compact() + "\n");
                _healthHandle.Write(line, 0, line.Length);
                _healthHandle.Flush(flushToDisk: true);
            }

            _health.Heartbeats += 1;
            switch (level)
            {
                case "green": _health.GreenSeconds += 1; break;
                case "amber": _health.AmberSeconds += 1; break;
                default: _health.RedSeconds += 1; break;
            }
            int Rank(string l) => l == "red" ? 2 : l == "amber" ? 1 : 0;
            if (Rank(level) > Rank(_health.WorstLevel)) _health.WorstLevel = level;
            _health.RecordsWritten += 1;
        }

        /// <summary>Drain remaining buffered audio, flush, and rewrite <c>session.json</c> as closed+populated.</summary>
        public void Close()
        {
            if (Closed) return;
            // Drain whatever is left in both rings before finalizing.
            for (int i = 0; i < 64; i++) { if (Pump() == 0) break; }
            _wav?.Close();
            int bytes = _wav?.TotalBytesOnDisk ?? 0;
            var part = new SessionContract.AudioPart(0, bytes, _firstAudioMs, _lastAudioMs);
            var parts = bytes > 0
                ? new List<SessionContract.AudioPart> { part }
                : new List<SessionContract.AudioPart>();
            WriteSessionJson("closed", _clock(), parts);
            _healthHandle?.Dispose();
            _healthHandle = null;
            _wav = null; // stream already disposed by _wav.Close() above
            Closed = true;
        }

        /// <summary>
        /// Deterministically release the WAV + health OS file handles even when <see cref="Close"/> was
        /// never reached — an abandoned/errored/never-closed session (e.g. capture fails to start after
        /// <see cref="Start"/>, or a call that captures nothing). Idempotent and safe to call after
        /// <see cref="Close"/>. Crucially it does NOT rewrite <c>session.json</c>: an un-closed session
        /// stays in its loud "open" state — the never-silent anomaly the pipeline's reconcile guard flags
        /// — rather than being silently marked "closed". Without this, on Windows the leaked WAV handle
        /// leaves <c>audio-part-00.wav</c> locked (a crashed/killed companion could not be cleaned up or
        /// re-collected); POSIX masked the leak by allowing unlink of an open file.
        /// </summary>
        public void Dispose()
        {
            _wav?.Dispose();
            _wav = null;
            _healthHandle?.Dispose();
            _healthHandle = null;
        }

        private void WriteSessionJson(string status, long? endedAt, IReadOnlyList<SessionContract.AudioPart> parts)
        {
            var json = SessionContract.SessionJson(Params, status, endedAt, parts, _health);
            string path = System.IO.Path.Combine(Dir, "session.json");
            File.WriteAllBytes(path, json.Data());
        }
    }
}

File: richos/tools/richos-service/companion-macos/Sources/RichOSCompanionCore/SessionWriter.swift
import Foundation

/// Orchestrates one capture session's on-disk contract directory (architecture §6.1 never-silent).
///
/// Lifecycle:
///   1. `start()`  — create the dir, write `session.json` (status "open", empty audio) IMMEDIATELY,
///                   open the WAV. A call that captures nothing therefore leaves a loud open/empty
///                   session on disk — an anomaly the pipeline's reconcile guard flags — never an
///                   absence nobody notices.
///   2. `pushMic` / `pushSystem` — real-time callbacks feed mono Float frames.
///   3. `pump()`   — mix aligned frames (with failover), append to WAV, accrue byte growth.
///   4. `writeHealthTick()` — one `health.ndjson` row per second.
///   5. `close()`  — flush + fsync WAV, rewrite `session.json` (status "closed", populated audio +
///                   health). The closed status is what triggers P1's pipeline.
public final class SessionWriter {
    public let dir: String
    public let params: SessionContract.Params
    private let clock: () -> Int64

    private var wav: WavWriter?
    private let micRing = RingBuffer()
    private let sysRing = RingBuffer()

    private var firstAudioMs: Int64?
    private var lastAudioMs: Int64?
    private var lastBytesTotal = 0
    private var health = SessionContract.HealthSummary()
    private var healthHandle: FileHandle?

    /// Failover threshold: a source silent this long while the other is live trips silence-fill+alarm.
    public var starveThresholdMs: Int64 = 15_000

    public private(set) var closed = false
    public private(set) var lastDecision: MixDecision?

    public init(rootZone: String, params: SessionContract.Params, clock: @escaping () -> Int64) {
        self.dir = (rootZone as NSString).appendingPathComponent(params.sessionId)
        self.params = params
        self.clock = clock
    }

    /// Write `session.json` (open) + open the WAV part. Idempotent-safe to call once.
    public func start() throws {
        try FileManager.default.createDirectory(atPath: dir, withIntermediateDirectories: true)
        try writeSessionJSON(status: "open", endedAt: nil, parts: [])
        let wavPath = (dir as NSString).appendingPathComponent(SessionContract.audioPartFile(0))
        wav = try WavWriter(path: wavPath, sampleRate: params.sampleRate, channels: 2, bitsPerSample: 16)
        let healthPath = (dir as NSString).appendingPathComponent("health.ndjson")
        FileManager.default.createFile(atPath: healthPath, contents: nil)
        healthHandle = FileHandle(forWritingAtPath: healthPath)
    }

    public func pushMic(_ mono: [Float]) { micRing.append(mono, nowMs: clock()) }
    public func pushSystem(_ mono: [Float]) { sysRing.append(mono, nowMs: clock()) }

    /// Mix one block of aligned frames to the WAV. Returns bytes written this pump.
    @discardableResult
    public func pump() throws -> Int {
        guard let wav = wav, !closed else { return 0 }
        let now = clock()
        let micSilent = micRing.lastAppend() == 0 ? 0 : now - micRing.lastAppend()
        let sysSilent = sysRing.lastAppend() == 0 ? 0 : now - sysRing.lastAppend()
        let decision = MixDecision.decide(
            micAvail: micRing.count,
            systemAvail: sysRing.count,
            micSilentMs: micSilent,
            systemSilentMs: sysSilent,
            starveThresholdMs: starveThresholdMs
        )
        lastDecision = decision
        guard decision.frames > 0 else { return 0 }

        let mic = decision.useMic ? micRing.take(decision.frames)
                                  : [Float](repeating: 0, count: decision.frames)
        let sys = decision.useSystem ? sysRing.take(decision.frames)
                                     : [Float](repeating: 0, count: decision.frames)
        // Guard alignment even under failover silence-fill.
        let n = min(mic.count, sys.count)
        let interleaved = ChannelMixer.interleaveToInt16(mic: Array(mic[0..<n]), system: Array(sys[0..<n]))
        let bytes = ChannelMixer.int16LEBytes(interleaved)
        try wav.append(bytes)
        if firstAudioMs == nil { firstAudioMs = now }
        lastAudioMs = now
        return bytes.count
    }

    /// Flush the WAV to disk (fsync) — the ≤ one-flush-window crash guarantee.
    public func flush() throws { try wav?.flush() }

    /// Total bytes on disk for the audio part so far — the watchdog's ground-truth liveness signal.
    public var pumpedBytesTotal: Int { wav?.totalBytesOnDisk ?? 0 }

    /// Emit one health row + accrue the rolled-up summary from precomputed per-channel levels
    /// (peak + RMS over the last second). Kept level-based (not block-based) so the live capture can
    /// meter cheaply in its audio callbacks without re-buffering a second of samples.
    public func writeHealthTick(
        micPeak: Double, micRms: Double, sysPeak: Double, sysRms: Double,
        tapRunning: Bool, micRunning: Bool
    ) throws {
        let now = clock()
        let bytesTotal = wav?.totalBytesOnDisk ?? 0
        let delta = bytesTotal - lastBytesTotal
        lastBytesTotal = bytesTotal

        var problems: [String] = []
        var level = "green"
        // Never-silent signals (§6.2): stalled byte growth, and failover-tripped starvation.
        if health.heartbeats > 20 && delta == 0 { problems.append("no audio bytes written in the last second"); level = "amber" }
        if let d = lastDecision, d.micStarved { problems.append("microphone source stalled — silence-filling LEFT (mic-only failover the other way)"); level = "red" }
        if let d = lastDecision, d.systemStarved { problems.append("system-audio tap stalled — silence-filling RIGHT (possible permission revocation)"); level = "red" }
        if !tapRunning { problems.append("system tap IOProc not running"); level = "red" }
        if !micRunning { problems.append("microphone input not running"); level = "red" }

        let sample = SessionContract.HealthSample(
            sessionId: params.sessionId, t: now,
            micRms: micPeak, sysRms: sysPeak,
            micRmsMean: micRms, sysRmsMean: sysRms,
            bytesTotal: bytesTotal, bytesDelta: delta, part: 0,
            tapRunning: tapRunning, micRunning: micRunning, level: level, problems: problems
        )
        if let h = healthHandle {
            try h.write(contentsOf: Data((sample.json().compact() + "\n").utf8))
        }
        health.heartbeats += 1
        switch level {
        case "green": health.greenSeconds += 1
        case "amber": health.amberSeconds += 1
        default: health.redSeconds += 1
        }
        let rank = ["green": 0, "amber": 1, "red": 2]
        if (rank[level] ?? 0) > (rank[health.worstLevel] ?? 0) { health.worstLevel = level }
        health.recordsWritten += 1
    }

    /// Drain remaining buffered audio, flush, and rewrite `session.json` as closed+populated.
    public func close() throws {
        guard !closed else { return }
        // Drain whatever is left in both rings before finalizing.
        for _ in 0..<64 { if try pump() == 0 { break } }
        try wav?.close()
        let bytes = wav?.totalBytesOnDisk ?? 0
        let part = SessionContract.AudioPart(part: 0, bytes: bytes, firstChunkAt: firstAudioMs, lastChunkAt: lastAudioMs)
        try writeSessionJSON(status: "closed", endedAt: clock(), parts: bytes > 0 ? [part] : [])
        try healthHandle?.close()
        closed = true
    }

    private func writeSessionJSON(status: String, endedAt: Int64?, parts: [SessionContract.AudioPart]) throws {
        let json = SessionContract.sessionJSON(params, status: status, endedAt: endedAt, audioParts: parts, health: health)
        let path = (dir as NSString).appendingPathComponent("session.json")
        try json.data().write(to: URL(fileURLWithPath: path))
    }
}
